src/main/java/com/ibm/cloud/sdk/core/security/TokenRequestBasedAuthenticator.java

```diff
@@ -211,12 +236,26 @@ public void authenticate(Builder builder) {
   public String getToken() {
     String token;
 
-    if (this.tokenData == null || !this.tokenData.isTokenValid()) {
-      // request new token
-      this.tokenData = requestToken();
+    if (this.tokenData == null) {
+      this.tokenData = synchronizedRequestToken();
+    } else if (this.tokenData.needsRefresh()) {
```

Comment:
Not to pile on here too much, but the logic might work out better if this second check was to detect that the tokenData represents a totally expired access token, in which case you'd just want to call synchronizedRequestToken() just like in the first check (for a null tokenData).
Then the third check would be to check if we need to refresh (even though the current token is valid but within the refresh window).   Just my $.02 :)

Reply:
I just pushed a commit that addresses Mike's latest comments and I think that should take care of your second point about the `getToken()` logic.

I also did tweak the `IamToken` class, but I didn't break it out like we do with CP4D, mostly because I was a bit worried about moving public methods on `IamToken` to get fields in case it was a breaking change. I added the calculation logic to the constructor and that's working the same.